daffodil-io/src/main/scala/org/apache/daffodil/processors/charset/CharsetUtils.scala

```diff
@@ -271,11 +271,7 @@ trait EncoderDecoderMixin
                 val truncString = cb.toString()
                 truncString
               }
-              case _ if cr.isMalformed() => {
-                cr.throwException()
-                ""
-              }
-              case _ if cr.isUnmappable() => {
+              case _ if cr.isMalformed() || cr.isUnmappable=> {
                 cr.throwException()
                 ""
```

Comment:
cr.throwException's type is wrong if these clauses need this "" at the end. cr.throwException() should have return type Nothing, which should be fine as the return type.

Reply:
I didn't find this to be the case. This match is used to build a string, so not having every case return a string leads to s Type Mismatch Error (Required String, Found Unit).

Reply:
The problem is that ``cr`` is a Java [CoderResult](https://docs.oracle.com/javase/7/docs/api/java/nio/charset/CoderResult.html), which just returns void, so Scala can't know that throwException will always throw and exception and whatever follows it doesn't matter. Returning an empty string is a codepath that is impossible to hit. Maybe instead of returning "" it should be Assert.impossible() or something? Assert.impossible does return Nothing so it should make the scala compiler happy.

Reply:
That did make the compiler happy!